packages/core/src/browser/quick-open/quick-open-service.ts

```diff
@@ -57,12 +63,257 @@ export namespace QuickOpenOptions {
+    }
+
+    getDefaultHandler(): QuickOpenHandler {
+        return this.defaultHandler;
+    }
+
+    isDefaultHandler(handler: QuickOpenHandler): boolean {
+        return handler === this.getDefaultHandler();
+    }
+
+    /**
+     * Return all registered handlers.
+     */
+    getHandlers(): QuickOpenHandler[] {
+        return [...this.handlers.values()];
+    }
+
+    /**
+     * Return a handler that matches the given text or the default handler if none.
+     */
+    getHandlerOrDefault(text: string): QuickOpenHandler {
+        for (const handler of this.handlers.values()) {
+            if (text.startsWith(handler.prefix)) {
+                return handler;
+            }
+        }
+        return this.getDefaultHandler();
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+}
+
+/**
+ * The quick open model just delegates all the calls to another model depending on the used prefix.
+ */
+class PrefixModel implements QuickOpenModel {
+
+    /**
+     * Stores the current handler for the quick open widget opened in a prefix-aware mode.
+     * Allows to detect that another model is called.
+     */
+    protected currentHandler: QuickOpenHandler;
+
+    constructor(
+        protected readonly handlers: QuickOpenHandlerRegistry,
+        protected readonly openService: QuickOpenService) {
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const handler = this.handlers.getHandlerOrDefault(lookFor);
+        if (handler !== this.currentHandler) {
+            this.onHandlerChanged(handler, lookFor);
+        } else {
+            const handlerModel = await handler.getModel();
+            const searchValue = this.handlers.isDefaultHandler(handler) ? lookFor : lookFor.substr(handler.prefix.length);
+            handlerModel.onType(searchValue, items => acceptor(items));
+        }
+    }
+
+    /**
+     * Handles changing the currently used handler as a result of user's typing.
+     * @param handler quick open handler that should be used
+     * @param prefix value that is currently typed in the quick open widget
+     */
+    protected onHandlerChanged(handler: QuickOpenHandler, prefix: string): void {
+        // need to notify the previous handler's model about closing the quick open widget
+        if (this.currentHandler) {
+            const closingHandler = this.currentHandler.getOptions().onClose;
+            if (closingHandler) {
+                closingHandler(true);
+            }
+        }
+        this.currentHandler = handler;
+        handler.init();
+        // a prefix has been changed, so the quick open widget needs to be reopened with the new options
+        this.openService.show(prefix);
+    }
+}
+
 @injectable()
 export class QuickOpenService {
+
     /**
+     * The internal model which is used when the quick open widget is opened in handling prefix mode.
+     */
+    protected model: QuickOpenModel;
```

Comment:
How about we built the prefix-based quick open service as a separate class on the top of `QuickOpenService`? This code is not needed for other quick-open services. I would move new APIs to a separate file to leave low-level API uncluttered.